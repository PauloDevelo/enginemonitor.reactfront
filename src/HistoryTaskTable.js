import React from 'react';
import { Table } from 'reactstrap';
import { FormattedMessage, FormattedDate } from 'react-intl';
import PropTypes from 'prop-types';
import tasktablemsg from "./TaskTable.messages";
import { shorten } from './TaskHelper'; 

import './HistoryTaskTable.css';

const TaskRow = ({entry, onClick}) => {
    var remarks = entry.remarks.replace(/\n/g, '<br />');
    var shortenRemarks = shorten(remarks);
    var ageStr = entry.age === -1?"":entry.age + 'h';
    var entryDate = new Date(entry.UTCDate)

    return(
        <tr className='small clickable' onClick={() => onClick()}>
            <td><FormattedDate value={entryDate} /></td>
            <td>{ageStr}</td>
            <td dangerouslySetInnerHTML={{ __html: shortenRemarks }}></td>
        </tr>
    );
}

export default function HistoryTaskTable({taskHistory, toggleEntryModal, classNames}){
    var history = [];
    if(taskHistory){
        history = taskHistory.map(entry => <TaskRow key={entry.id} entry={entry} onClick={() => toggleEntryModal(false, entry)}/>);
        history.reverse();
    }

    return(
        <div className={classNames}>
            <Table responsive size="sm" hover striped>
                <thead className="thead-light">
                    <tr>
                        <th><FormattedMessage {...tasktablemsg.ackDate} /></th>
                        <th><FormattedMessage {...tasktablemsg.engineAge} /></th>
                        <th><FormattedMessage {...tasktablemsg.remarks} /></th>
                    </tr>
                </thead>
                <tbody>
                    {history}
                </tbody>
            </Table>
        </div>
    );
}

HistoryTaskTable.propTypes = {
    taskHistory: PropTypes.array.isRequired,
    toggleEntryModal: PropTypes.func.isRequired,
    classNames: PropTypes.string
};