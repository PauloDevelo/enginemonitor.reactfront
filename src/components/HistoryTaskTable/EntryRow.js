import React from 'react';
import { FormattedDate } from 'react-intl';
import PropTypes from 'prop-types';

import { shorten } from '../../helpers/TaskHelper'; 

export default function EntryRow({entry, onClick}){
    var remarks = entry.remarks.replace(/\n/g, '<br />');
    var shortenRemarks = shorten(remarks);
    var ageStr = entry.age === -1?"":entry.age + 'h';
    var entryDate = new Date(entry.date)

    return(
        <tr className='small clickable' onClick={() => onClick()}>
            <td><FormattedDate value={entryDate} /></td>
            <td>{ageStr}</td>
            <td dangerouslySetInnerHTML={{ __html: shortenRemarks }}></td>
        </tr>
    );
}

EntryRow.propTypes = {
    entry: PropTypes.object.isRequired,
    onClick: PropTypes.func.isRequired
};