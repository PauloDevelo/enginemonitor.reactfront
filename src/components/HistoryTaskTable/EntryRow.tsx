import React from 'react';
import { FormattedDate } from 'react-intl';
import PropTypes from 'prop-types';

import { shorten } from '../../helpers/TaskHelper';
import { Entry } from '../../types/Types';

type Props = {
    entry: Entry,
    onClick: ()=>void
};

export default function EntryRow({entry, onClick}: Props){
    var remarks = entry.remarks.replace(/\n/g, '<br />');
    var shortenRemarks = shorten(remarks);
    var ageStr = entry.age === -1?"":entry.age + 'h';
    var entryDate = new Date(entry.date)

    return(
        <tr className='small clickable' onClick={onClick}>
            <td><FormattedDate value={entryDate} /></td>
            <td>{ageStr}</td>
            <td dangerouslySetInnerHTML={{ __html: shortenRemarks }}></td>
        </tr>
    );
}

EntryRow.propTypes = {
    entry: PropTypes.object.isRequired,
    onClick: PropTypes.func.isRequired
};