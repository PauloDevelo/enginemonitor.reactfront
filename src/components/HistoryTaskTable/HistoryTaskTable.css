.historytasktable{
  display: flex;
  flex-direction: column;
}