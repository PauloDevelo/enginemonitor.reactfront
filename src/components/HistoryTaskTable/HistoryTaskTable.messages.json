{
    "ackDate": {
        "id": "ackDate",
        "defaultMessage": "Realisation date",
        "description": "Colum header"
    },
    "age": {
		"id": "ageColumHeader",
        "defaultMessage": "Equipment Age",
		"description": "Colum header"
    },
    "remarks": {
        "id": "remarks",
        "defaultMessage": "Remarks",
        "description": "Colum header"
    },
    "errorFetching": {
        "id": "errorFetching",
        "defaultMessage": "Something went wrong ...",
        "description": "Error message"
    },
    "taskHistoryTitle": {
        "id": "taskHistoryTitle",
        "defaultMessage": "Task history",
        "description": "Title"
    },
    "monthperiod": {
        "id": "monthperiod",  
        "defaultMessage": "{month, plural, one {month} other {# months}}",
        "description": "Month period"
    },
    "yearperiod": {
        "id": "yearperiod",  
        "defaultMessage": "{year, plural, one {year} other {# years}}",
        "description": "Year period"
    },
    "dayperiod": {
        "id": "dayperiod",  
        "defaultMessage": "{day, plural, one {day} other {# days}}",
        "description": "Day period"
    }
}
