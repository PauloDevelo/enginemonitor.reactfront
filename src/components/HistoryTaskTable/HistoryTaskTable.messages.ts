import {defineMessages, Messages} from "react-intl";

const messages: Messages = {
    ackDate: {
        id: "ackDate",
        defaultMessage: "Realisation date",
        description: "Colum header"
    },
    age: {
		id: "age",
        defaultMessage: "Equipment Age",
		description: "Colum header"
    },
    remarks: {
        id: "remarks",
        defaultMessage: "Remarks",
        description: "Colum header"
    },
    errorFetching: {
        id: "errorFetching",
        defaultMessage: "Something went wrong ...",
        description: "Error message"
    }
};

const definedMessages: typeof messages = defineMessages(messages);

export default definedMessages;