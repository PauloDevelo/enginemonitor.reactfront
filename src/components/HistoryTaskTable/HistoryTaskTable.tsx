import  React, { useEffect, useState, Fragment } from 'react';
import { Button } from 'reactstrap';
import { 
    composeDecorators,
    withInMemorySortingContext,
    withFixedHeader,
    withHeaderControl,
  } from 'react-table-factory';
import { defineMessages, Messages, FormattedMessage, FormattedDate } from 'react-intl';
import ModalEditEntry from '../ModalEditEntry/ModalEditEntry';
import Loading from '../Loading/Loading';

import { faCheckSquare, faExclamationTriangle } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import PropTypes from 'prop-types';

import * as moment from 'moment';
import EquipmentMonitorService from '../../services/EquipmentMonitorServiceProxy';

import { useEditModal } from '../../hooks/EditModalHook';

import { shorten } from '../../helpers/TaskHelper';
import { createDefaultEntry } from '../../helpers/EntryHelper';

import jsonMessages from "./HistoryTaskTable.messages.json";
const messages: Messages = defineMessages(jsonMessages);

import './HistoryTaskTable.css';

import { Equipment, Task, Entry, AgeAcquisitionType } from '../../types/Types';

type Props = {
    equipment: Equipment | undefined,
    task: Task | undefined, 
    taskHistoryRefreshId: number,
    onHistoryChanged: (newEntries: Entry[])=>void,
    classNames: string
}

const Table = composeDecorators(
  withHeaderControl,
  withInMemorySortingContext(),
  withFixedHeader // should be last
)();

enum FetchState{
    StandBy,
    Fetching,
    Error
}

const HistoryTaskTable = ({equipment, task, taskHistoryRefreshId, onHistoryChanged, classNames}: Props) => {
    const equipmentId = equipment ? equipment._id : undefined;
    const taskId = task ? task._id : undefined;

    const modalHook = useEditModal<Entry | undefined>(undefined);

    const [entries, setEntries] = useState<Entry[]>([]);
    const [fetchingState, setFetchingState] = useState(FetchState.StandBy);

    useEffect(() => {
      fetchEntries();
    }, [taskId, taskHistoryRefreshId]);

    const fetchEntries = async () => {
        setFetchingState(FetchState.Fetching);

        try {
            let newEntries:Entry[] = [];
            if(equipmentId && taskId){
                newEntries = await EquipmentMonitorService.fetchEntries(equipmentId, taskId);
            }
            
            setEntries(newEntries);
            setFetchingState(FetchState.StandBy);
        } 
        catch (error) {
            setFetchingState(FetchState.Error);
        }
    };

    const changeEntries = (newEntries: Entry[]) => {
      setEntries(newEntries);
      if(onHistoryChanged){
        onHistoryChanged(newEntries);
      }
    };

    const onSavedEntry = (savedEntry: Entry) => {
        const newCurrentHistoryTask = entries.filter(entry => entry._id !== savedEntry._id);
        newCurrentHistoryTask.unshift(savedEntry);
        newCurrentHistoryTask.sort((entryA, entryB) => entryA.date.getTime() - entryB.date.getTime());

        changeEntries(newCurrentHistoryTask);
	}
	
	const onDeleteEntry = async(entry: Entry) => {  
        var newCurrentHistoryTask = entries.slice(0).filter(e => e._id !== entry._id);
        changeEntries(newCurrentHistoryTask);
    }

    const innerEntryCell = (entry:Entry, content: JSX.Element, classNames?: string) => {
		classNames = classNames === undefined ? '' : classNames;
		return (
			<div onClick={() => modalHook.displayData(entry)} className={classNames + ' innerTd clickable'} >
				{content}
			</div>
		);
	}

    const columns = [
		{
			name: 'date',
			header: () => (
				<div className={'innerTdHead'}>
					<FormattedMessage {...messages.ackDate} />
				</div>
			),
			cell: (content: any) => {
                const entryDate = new Date(content.data.date);
				return innerEntryCell(content.data, <FormattedDate value={entryDate} />);
            },
            style: {width:"25%"},
			sortable: true
		},
		{
			name: 'age',
			header: () => (
				<div className={'innerTdHead'}>
					<FormattedMessage {...messages.age} />
				</div>
			),
			cell: (content: any) => {
                const entry:Entry = content.data;
                const equ = equipment as Equipment;

                if (equ.ageAcquisitionType !== AgeAcquisitionType.time){
                    return innerEntryCell(entry, <Fragment>{entry.age === -1?"":entry.age + 'h'}</Fragment>);
                }
                else{
                    const diff = moment.duration(entry.date.getTime() - equ.installation.getTime());
                    const year = diff.years();
                    const month = diff.months();
                    const day = diff.days();
                    return innerEntryCell(entry, <Fragment>
                        {diff.years() > 0 && <FormattedMessage {... messages.yearperiod} values={{year}}/>}{' '}
                        {diff.months() > 0 && <FormattedMessage {... messages.monthperiod} values={{month}}/>}{' '}
                        {diff.days() > 0 && <FormattedMessage {... messages.dayperiod} values={{day}}/>}
                    </Fragment>);
                }
            },
            style: {width:"25%"},
			sortable: true
		},
		{
			name: 'remarks',
			header: () => (
				<div className={'text-center innerTdHead'}>
					<FormattedMessage {...messages.remarks}/>
				</div>
			),
			cell: (content: any) => {
                const entry:Entry = content.data;
                var remarks = entry.remarks.replace(/\n/g, '<br />');
                var shortenRemarks = shorten(remarks);

				return innerEntryCell(entry, <div dangerouslySetInnerHTML={{ __html: shortenRemarks }}/>);
            },
            style: {width:"50%"},
			sortable: false
		}
	];
    
    return(
        <div className={classNames + ' historytasktable'}>
            <span className="mb-2">
                <b><FormattedMessage {...messages.taskHistoryTitle} /></b>
                {task && <Button aria-label="Add" color="success" size="sm" className="float-right mb-2" onClick={() => {
                    modalHook.displayData(createDefaultEntry(equipment, task));
                }}>
                    <FontAwesomeIcon icon={faCheckSquare} />
                </Button>}
            </span>
            {fetchingState === FetchState.Error && <div><FontAwesomeIcon icon={faExclamationTriangle} color="red"/><FormattedMessage {...messages.errorFetching} /></div>}
            {fetchingState === FetchState.Fetching && <Loading/>}
            {fetchingState === FetchState.StandBy && 
            <Table
                data={entries}
                className="default-theme"
                defaultSortParameter="date"
                defaultSortDirection="desc"
                columns={columns}
            />
            }
            
            {equipment && task && modalHook.data && <ModalEditEntry 
                equipment={equipment}
                task={task}
                entry={modalHook.data}
                saveEntry={onSavedEntry} 
                deleteEntry={onDeleteEntry}
                visible={modalHook.editModalVisibility}
                toggle={modalHook.toggleModal}
                className='modal-dialog-centered'
            />}
        </div>
    );
}

HistoryTaskTable.propTypes = {
    equipment: PropTypes.object,
    task: PropTypes.object,
    classNames: PropTypes.string,
    onHistoryChanged: PropTypes.func,
};

export default React.memo(HistoryTaskTable);