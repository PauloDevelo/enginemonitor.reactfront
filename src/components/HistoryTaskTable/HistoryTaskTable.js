import  React, { useState, useEffect } from 'react';
import { Table, Button } from 'reactstrap';
import { FormattedMessage } from 'react-intl';
import { faCheckSquare, faExclamationTriangle } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import {CSSTransition, TransitionGroup} from 'react-transition-group'
import PropTypes from 'prop-types';

import EquipmentMonitorService from '../../services/EquipmentMonitorServiceProxy';
import ModalEditEntry from '../ModalEditEntry/ModalEditEntry';
import EntryRow from './EntryRow';
import Loading from '../Loading/Loading'

import { createDefaultEntry } from '../../helpers/EntryHelper'

import taskTableMsg from "../TaskTable/TaskTable.messages";

import './HistoryTaskTable.css';
import '../../style/transition.css';

const HistoryTaskTable = ({equipment, task, onHistoryChanged, classNames}) => {
    const [editEntryModalVisibility, setEditEntryModalVisibility] = useState(false);
    const [editedEntry, setEditedEntry] = useState(undefined);
    const [taskHistory, setTaskHistory] = useState([]);
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);

    const toggleEditEntryModal = () => {
        setEditEntryModalVisibility(!editEntryModalVisibility);
    }

    const fetchEntries = async() => {
        setIsError(false);
        setIsLoading(true);

        try{
            let entries = [];
            if(equipment && task){
                entries = await EquipmentMonitorService.fetchEntries(equipment._id, task._id);
            }
            setTaskHistory(entries);
        }
        catch(error){
            setIsError(true);
        }

        setIsLoading(false);
    }

    useEffect(() => {
        fetchEntries();
    }, [task]);

    const onSavedEntry = (savedEntry) => {
            const newCurrentHistoryTask = taskHistory.filter(entry => entry._id !== savedEntry._id);
            newCurrentHistoryTask.unshift(savedEntry);
            newCurrentHistoryTask.sort((entrya, entryb) => { return entrya.date - entryb.date; });

            changeCurrentHistory(newCurrentHistoryTask);
	}
	
	const onDeleteEntry = async(entryId) => {  
            var newCurrentHistoryTask = taskHistory.slice(0).filter(e => e._id !== entryId);

            changeCurrentHistory(newCurrentHistoryTask);
    }
    
    const changeCurrentHistory = (newCurrentHistoryTask) => {
        setTaskHistory(newCurrentHistoryTask);

        if(onHistoryChanged){
            onHistoryChanged(newCurrentHistoryTask);
        }
    }

    let history = [];
    if(taskHistory && taskHistory.length > 0){
        history = taskHistory.map(entry => {
        return(
            <CSSTransition key={entry._id} in={true} timeout={500} classNames="tr">
                <EntryRow entry={entry} onClick={() => {
                    setEditedEntry(entry);
                    setEditEntryModalVisibility(true);
                }}/>
            </CSSTransition>
            )}
        );
        history.reverse();
    }

    return(
        <div className={classNames}>

            <span className="mb-2">
                <Button color="success" size="sm" className="float-right mb-2" onClick={() => {
                    setEditedEntry(createDefaultEntry(equipment, task));
                    setEditEntryModalVisibility(true);
                }}>
                    <FontAwesomeIcon icon={faCheckSquare} />
                </Button>
            </span>
            {isError && <div><FontAwesomeIcon icon={faExclamationTriangle} color="red"/><FormattedMessage {...taskTableMsg.errorFetching} /></div>}
            {isLoading ? !isError && <Loading/> :
            !isError && <Table responsive size="sm" hover striped>
                <thead className="thead-light">
                    <tr>
                        <th><FormattedMessage {...taskTableMsg.ackDate} /></th>
                        <th><FormattedMessage {...taskTableMsg.age} /></th>
                        <th><FormattedMessage {...taskTableMsg.remarks} /></th>
                    </tr>
                </thead>
                <TransitionGroup component="tbody">
                    {history}
                </TransitionGroup>
            </Table>
            }
            

            <ModalEditEntry 
                equipment={equipment}
                task={task}
                entry={editedEntry}
                saveEntry={onSavedEntry} 
                deleteEntry={onDeleteEntry}
                visible={editEntryModalVisibility}
                toggle={toggleEditEntryModal}
                className='modal-dialog-centered'
            />
        </div>
    );
}

HistoryTaskTable.propTypes = {
    equipment: PropTypes.object,
    task: PropTypes.object,
    classNames: PropTypes.string,
    onHistoryChanged: PropTypes.func,
};

export default HistoryTaskTable;