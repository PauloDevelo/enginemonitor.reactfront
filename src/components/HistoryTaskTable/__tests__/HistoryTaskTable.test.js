import ignoredMessages from '../../../testHelpers/MockConsole';
import chai, {assert} from 'chai';

import React from 'react';
import { IntlProvider } from "react-intl";

import { mount } from 'enzyme';

import localforage from 'localforage';
import imageProxy from '../../../services/ImageProxy';
import entryProxy from '../../../services/EntryProxy';

import {AgeAcquisitionType} from '../../../types/Types';
import HistoryTaskTable from '../HistoryTaskTable';
import updateWrapper from '../../../testHelpers/EnzymeHelper';

jest.mock('../../../services/ImageProxy');
jest.mock('../../../services/EntryProxy');
jest.mock('localforage');

chai.use(require('chai-datetime'));

describe("HistoryTaskTable", () => {
    const equipment = {
        _uiId: "equipment_01",
		name: "moteur",
		brand: "Nanni",
		model: "N3.30",
		age: 2750,
		installation: new Date("2019-11-07T23:39:36.288Z"),
		ageAcquisitionType: AgeAcquisitionType.manualEntry,
		ageUrl: ""
    };

    const task = {
        _uiId: "task_01",
		name: 'vidange',
		usagePeriodInHour: 200,
		periodInMonth: 6,
        description: 'Change the oil',
        nextDueDate: new Date("2019-11-07T23:39:36.288Z"),
        level: 0,
        usageInHourLeft: 100,
	}

    const entries = [
        {
            _uiId: "entry_01",
            name: "vidange",
            date: new Date("2019-11-08T00:11:18.112Z"),
            age: 1234,
            remarks: 'RAS',
            taskUiId: "task_01",
            equipmentUiId: "equipment_01"
        },
        {
            _uiId: "entry_02",
            name: "vidange inverseur",
            date: new Date("2019-11-06T00:11:18.112Z"),
            age: 1214,
            remarks: 'RAS',
            taskUiId: "task_01",
            equipmentUiId: "equipment_01"
        },
        {
            _uiId: "entry_03",
            name: "vidange inverseur",
            date: new Date("2019-11-09T00:11:18.112Z"),
            age: 1214,
            remarks: 'RAS',
            taskUiId: "task_01",
            equipmentUiId: "equipment_01"
        },
        {
            _uiId: "entry_04",
            name: "vidange inverseur",
            date: new Date("2019-11-09T00:11:19.112Z"),
            age: 1214,
            remarks: 'RAS',
            taskUiId: "task_01",
            equipmentUiId: "equipment_01"
        }
    ];

    beforeAll(() => {
        ignoredMessages.length = 0;
        ignoredMessages.push("test was not wrapped in act(...)");
        ignoredMessages.push("[React Intl] Missing message");
    });

    beforeEach(() => {
        jest.spyOn(entryProxy, "fetchEntries").mockImplementation(async (props) => {
            if (props.taskId === "task_01"){
                return Promise.resolve(entries);
            }
            Promise.resolve([]);
        });

        jest.spyOn(entryProxy, "existEntry").mockImplementation(async (equipmentId, entryId) => {
            if (entries.findIndex(e => e._uiId === entryId) !== -1){
                return Promise.resolve(true);
            }
            return Promise.resolve(false);
        });


        imageProxy.fetchImages.mockResolvedValue([]);
    });

    afterEach(() => {
        entryProxy.fetchEntries.mockRestore();
        entryProxy.existEntry.mockRestore();
        imageProxy.fetchImages.mockRestore();
    });

    it("Should render render the loading spinner while loading the entries", async() => {
        // Arrange
        const onHistoryChanged = jest.fn();
    
        // Act
        const historyTaskTable = mount(
            <IntlProvider locale={navigator.language}>
                <HistoryTaskTable 
                equipment={equipment} 
                task={task}
                onHistoryChanged={onHistoryChanged} 
                taskHistoryRefreshId={0}/>
            </IntlProvider>);
        
        // Assert
        expect(historyTaskTable.find('tbody').length).toBe(0);
        expect(historyTaskTable.find('Loading').length).toBe(1);
        expect(entryProxy.fetchEntries).toBeCalledTimes(1);
        expect(onHistoryChanged).toBeCalledTimes(0);

        expect(historyTaskTable).toMatchSnapshot();
    });

    it("Should render all the entries for a task", async() => {
        // Arrange
        const onHistoryChanged = jest.fn();
    
        // Act
        const historyTaskTable = mount(
            <IntlProvider locale={navigator.language}>
                <HistoryTaskTable 
                equipment={equipment} 
                task={task}
                onHistoryChanged={onHistoryChanged} 
                taskHistoryRefreshId={0}/>
            </IntlProvider>);
        await updateWrapper(historyTaskTable);

        // Assert
        const tbodyProps = historyTaskTable.find('tbody').at(0).props();
        for(let i = 0; i < entries.length; i++){
            expect(tbodyProps.children[i].props.data).toBe(entries[i]);
        }
        expect(entryProxy.fetchEntries).toBeCalledTimes(1);
        expect(onHistoryChanged).toBeCalledTimes(0);
        
        expect(onHistoryChanged).toMatchSnapshot();
    });

    it("Should render an empty table even when the task is undefined", async() => {
        // Arrange
        const onHistoryChanged = jest.fn();

        // Act
        const historyTaskTable = mount(
            <IntlProvider locale={navigator.language}>
                <HistoryTaskTable 
                equipment={equipment} 
                task={undefined}
                onHistoryChanged={onHistoryChanged} 
                taskHistoryRefreshId={0}/>
            </IntlProvider>);
        await updateWrapper(historyTaskTable);

        // Assert
        const tbodyProps = historyTaskTable.find('tbody').at(0).props();
        expect(tbodyProps.children.length).toBe(0);
        expect(entryProxy.fetchEntries).toBeCalledTimes(1);
        expect(onHistoryChanged).toBeCalledTimes(0);
        
        expect(onHistoryChanged).toMatchSnapshot();
    });

    it("Should rerender all the entries for a task when equipmentHistoryRefreshId change", async() => {
        // Arrange
        const onHistoryChanged = jest.fn();

        const properties = {equipment, task, onHistoryChanged, taskHistoryRefreshId: 0};
        const wrapper = mount(
            React.createElement(
                props => (
                    <IntlProvider locale={navigator.language}>
                        <HistoryTaskTable 
                        {...props}/>
                    </IntlProvider>
                ),
                properties)
            );
        await updateWrapper(wrapper);

        // Act
        properties.taskHistoryRefreshId = 1;

        wrapper.setProps(properties); 
        await updateWrapper(wrapper);
        
        // Assert
        expect(entryProxy.fetchEntries).toBeCalledTimes(2);
        expect(onHistoryChanged).toBeCalledTimes(0);
    });

    // it("Should open the edition entry modal when clicking on any cell", async() => {
    //     // Arrange
    //     const onEquipmentChanged = jest.fn();

    //     const equipmentHistoryTable = mount(
    //         <IntlProvider locale={navigator.language}>
    //             <EquipmentHistoryTable 
    //             equipment={equipment} 
    //             onTaskChanged={onEquipmentChanged} 
    //             equipmentHistoryRefreshId={0}/>
    //         </IntlProvider>);
    //     await updateWrapper(equipmentHistoryTable);
    //     const cells = equipmentHistoryTable.find('ClickableCell');

    //     for(let i = 0; i < entries.length; i++){
    //         for(let numColumn = 0; numColumn < 4; numColumn++){
    //             // Act
    //             cells.at(i * 4 + numColumn).simulate('click');
    //             await updateWrapper(equipmentHistoryTable);

    //             // Assert
    //             const editEntryModal = equipmentHistoryTable.find('ModalEditEntry');
    //             expect(editEntryModal.length).toBe(1);
    //             expect(editEntryModal.props().visible).toBe(true);
    //             expect(editEntryModal.props().equipment).toBe(equipment);
    //             expect(editEntryModal.props().entry).toBe(entries[i]);
    //         }
    //     }
    // });

    // it("Should display the entries from the most recent to the oldest by default", async() => {
    //     // Arrange
    //     const onEquipmentChanged = jest.fn();

    //     // Act
    //     const equipmentHistoryTable = mount(
    //         <IntlProvider locale={navigator.language}>
    //             <EquipmentHistoryTable 
    //             equipment={equipment} 
    //             onTaskChanged={onEquipmentChanged} 
    //             equipmentHistoryRefreshId={0}/>
    //         </IntlProvider>);
    //     await updateWrapper(equipmentHistoryTable);

    //     // Assert
    //     assertTableSortedByDate(equipmentHistoryTable);
    // });

    // it("Should add an entry and it should remain sorted by date", async() => {
    //     // Arrange
    //     const onEquipmentChanged = jest.fn();
        
    //     const equipmentHistoryTable = mount(
    //         <IntlProvider locale={navigator.language}>
    //             <EquipmentHistoryTable 
    //             equipment={equipment} 
    //             onTaskChanged={onEquipmentChanged} 
    //             equipmentHistoryRefreshId={0}/>
    //         </IntlProvider>);
    //     await updateWrapper(equipmentHistoryTable);

    //     const addButton = equipmentHistoryTable.find('Button');
    //     addButton.simulate('click');
    //     await updateWrapper(equipmentHistoryTable);

    //     const editEntryModal = equipmentHistoryTable.find('ModalEditEntry');
    //     const saveEntry = editEntryModal.props().saveEntry;

    //     const newEntry = {
    //         _uiId: "entry_088",
    //         name: "remplacement silent bloc",
    //         date: new Date("2019-09-08T00:11:18.112Z"),
    //         age: 125894,
    //         remarks: 'RAS',
    //         taskUiId: "task_08",
    //         equipmentUiId: "equipment_01"
    //     }
    //     // Act
    //     saveEntry(newEntry);
    //     await updateWrapper(equipmentHistoryTable);

    //     // Assert
    //     expect(editEntryModal.length).toBe(1);
    //     expect(editEntryModal.props().visible).toBe(true);
    //     expect(editEntryModal.props().equipment).toBe(equipment);

    //     expect(equipmentHistoryTable.find('ClickableCell').length).toBe((entries.length + 1) * 4);
    //     const newCells = equipmentHistoryTable.find('ClickableCell').findWhere(n => n.props().data === newEntry);
    //     expect(newCells.length).toBe(4);

    //     assertTableSortedByDate(equipmentHistoryTable);
    //     expect(onEquipmentChanged).toBeCalledTimes(1);
    // });

    // it("Should remove an entry", async() => {
    //     // Arrange
    //     const onTaskChanged = jest.fn();
        
    //     const equipmentHistoryTable = mount(
    //         <IntlProvider locale={navigator.language}>
    //             <EquipmentHistoryTable 
    //             equipment={equipment} 
    //             onTaskChanged={onTaskChanged} 
    //             equipmentHistoryRefreshId={0}/>
    //         </IntlProvider>);
    //     await updateWrapper(equipmentHistoryTable);

    //     const cells = equipmentHistoryTable.find('ClickableCell').at(5);
    //     cells.simulate('click');
    //     const entryToDelete = cells.props().data;
    //     await updateWrapper(equipmentHistoryTable);

    //     const editEntryModal = equipmentHistoryTable.find('ModalEditEntry');
    //     const deleteEntry = editEntryModal.props().deleteEntry;

    //     // Act
    //     deleteEntry(entryToDelete);
    //     await updateWrapper(equipmentHistoryTable);

    //     // Assert
    //     expect(editEntryModal.length).toBe(1);
    //     expect(editEntryModal.props().visible).toBe(true);
    //     expect(editEntryModal.props().equipment).toBe(equipment);
    //     expect(editEntryModal.props().entry).toBe(entryToDelete);

    //     expect(equipmentHistoryTable.find('ClickableCell').length).toBe((entries.length - 1) * 4);
    //     const newCells = equipmentHistoryTable.find('ClickableCell').findWhere(n => n.props().data === entryToDelete);
    //     expect(newCells.length).toBe(0);

    //     expect(onTaskChanged).toBeCalledTimes(1);
    // });

    // function assertTableSortedByDate(table){
    //     const cells = table.find('ClickableCell');
    //     let previousDate = undefined;
    //     for(let i = 0; i < entries.length; i++){
    //         for(let numColumn = 0; numColumn < 4; numColumn++){
    //             const currentDate = cells.at(i * 4 + numColumn).props().data.date;
                
    //             if(previousDate !== undefined && currentDate !== previousDate){
    //                 assert.afterTime(previousDate, currentDate);
    //             }

    //             previousDate = currentDate;
    //         }
    //     }
    // }
});