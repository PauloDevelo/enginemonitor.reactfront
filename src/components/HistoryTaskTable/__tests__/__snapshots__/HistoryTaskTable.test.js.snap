// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryTaskTable Should render all the entries with the correct classnames for a task 1`] = `[MockFunction]`;

exports[`HistoryTaskTable Should render an empty table even when the task is undefined 1`] = `[MockFunction]`;

exports[`HistoryTaskTable Should render render the loading spinner while loading the entries 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en-US"
  messages={Object {}}
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <Memo(HistoryTaskTable)
    equipment={
      Object {
        "_uiId": "equipment_01",
        "age": 2750,
        "ageAcquisitionType": 1,
        "ageUrl": "",
        "brand": "Nanni",
        "installation": 2019-11-07T23:39:36.288Z,
        "model": "N3.30",
        "name": "moteur",
      }
    }
    onHistoryChanged={[MockFunction]}
    task={
      Object {
        "_uiId": "task_01",
        "description": "Change the oil",
        "level": 0,
        "name": "vidange",
        "nextDueDate": 2019-11-07T23:39:36.288Z,
        "periodInMonth": 6,
        "usageInHourLeft": 100,
        "usagePeriodInHour": 200,
      }
    }
    taskHistoryRefreshId={0}
  >
    <div
      className="undefined historytasktable"
    >
      <span
        className="mb-2"
      >
        <b>
          <FormattedMessage
            defaultMessage="Task history"
            description="Title"
            id="taskHistoryTitle"
            values={Object {}}
          >
            Task history
          </FormattedMessage>
        </b>
        <Button
          aria-label="Add"
          className="float-right mb-2"
          color="success"
          onClick={[Function]}
          size="sm"
          tag="button"
        >
          <button
            aria-label="Add"
            className="float-right mb-2 btn btn-success btn-sm"
            onClick={[Function]}
            type="button"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    448,
                    512,
                    Array [],
                    "f14a",
                    "M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zm-204.686-98.059l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.248-16.379-6.249-22.628 0L184 302.745l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.25 16.379 6.25 22.628.001z",
                  ],
                  "iconName": "check-square",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-check-square fa-w-14 "
                data-icon="check-square"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zm-204.686-98.059l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.248-16.379-6.249-22.628 0L184 302.745l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.25 16.379 6.25 22.628.001z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </button>
        </Button>
      </span>
      <Loading>
        <div
          className="lds-spinner"
        >
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
          <div />
        </div>
      </Loading>
    </div>
  </Memo(HistoryTaskTable)>
</IntlProvider>
`;
